File: scripts/playwright-github-search.js
#!/usr/bin/env node

/**
 * Playwright script to search for GitHub-related presentation topics
 * on the DevDays Asia 2025 seminar website
 * 
 * This script demonstrates how to use Playwright for web scraping
 * to find GitHub-related content on seminar websites.
 */

const { chromium } = require('playwright');

async function searchGitHubTopics() {
    const browser = await chromium.launch({ headless: true });
    const context = await browser.newContext();
    const page = await context.newPage();

    try {
        console.log('🚀 Navigating to DevDays Asia 2025 seminar website...');
        await page.goto('https://www.digitimes.com.tw/Seminar/DevDaysAsia2025/index.html', {
            waitUntil: 'domcontentloaded',
            timeout: 30000
        });

        console.log('📋 Searching for GitHub-related topics...');
        
        // Search for various GitHub-related keywords in the page content
        const githubKeywords = ['github', 'git', 'repository', 'version control', 'copilot', 'ci/cd', 'devops'];
        const foundTopics = [];

        // Look for session titles, speaker names, and descriptions
        const sessionSelectors = [
            'h1, h2, h3, h4', // Session titles
            '.session-title, .talk-title, .presentation-title', // Common class names for session titles
            '.speaker-bio, .session-description, .talk-description', // Session descriptions
            'p, div, span' // General text content
        ];

        for (const selector of sessionSelectors) {
            try {
                const elements = await page.$$(selector);
                
                for (const element of elements) {
                    const text = await element.textContent();
                    if (text) {
                        const lowerText = text.toLowerCase();
                        
                        // Check if any GitHub-related keywords are found
                        for (const keyword of githubKeywords) {
                            if (lowerText.includes(keyword)) {
                                foundTopics.push({
                                    keyword: keyword,
                                    text: text.trim(),
                                    element: selector
                                });
                                break;
                            }
                        }
                    }
                }
            } catch (error) {
                // Skip selectors that don't exist
                continue;
            }
        }

        // Look for schedule or agenda sections
        console.log('📅 Checking schedule and agenda sections...');
        const scheduleSelectors = [
            '.schedule, .agenda, .program',
            '[id*="schedule"], [id*="agenda"], [id*="program"]',
            '[class*="schedule"], [class*="agenda"], [class*="program"]'
        ];

        for (const selector of scheduleSelectors) {
            try {
                const scheduleSection = await page.$(selector);
                if (scheduleSection) {
                    const scheduleText = await scheduleSection.textContent();
                    if (scheduleText) {
                        const lowerText = scheduleText.toLowerCase();
                        
                        for (const keyword of githubKeywords) {
                            if (lowerText.includes(keyword)) {
                                foundTopics.push({
                                    keyword: keyword,
                                    text: 'Found in schedule section',
                                    element: selector,
                                    fullText: scheduleText.substring(0, 200) + '...'
                                });
                            }
                        }
                    }
                }
            } catch (error) {
                continue;
            }
        }

        // Look for speaker information
        console.log('👥 Searching speaker information...');
        const speakerSelectors = [
            '.speaker, .presenter',
            '[class*="speaker"], [class*="presenter"]'
        ];

        for (const selector of speakerSelectors) {
            try {
                const speakers = await page.$$(selector);
                
                for (const speaker of speakers) {
                    const speakerText = await speaker.textContent();
                    if (speakerText) {
                        const lowerText = speakerText.toLowerCase();
                        
                        for (const keyword of githubKeywords) {
                            if (lowerText.includes(keyword)) {
                                foundTopics.push({
                                    keyword: keyword,
                                    text: speakerText.trim(),
                                    element: 'speaker-info'
                                });
                                break;
                            }
                        }
                    }
                }
            } catch (error) {
                continue;
            }
        }

        // Take a screenshot for documentation
        await page.screenshot({ 
            path: 'devdays-asia-2025-screenshot.png',
            fullPage: true 
        });

        console.log('\n🎯 GitHub-related topics found:');
        console.log('=====================================');
        
        if (foundTopics.length === 0) {
            console.log('❌ No GitHub-related topics found on this page.');
            console.log('💡 This could mean:');
            console.log('   - The content is loaded dynamically (requires waiting)');
            console.log('   - GitHub topics are in different sections not yet searched');
            console.log('   - The seminar focuses on other technologies');
        } else {
            // Remove duplicates and display results
            const uniqueTopics = foundTopics.reduce((acc, current) => {
                const exists = acc.find(item => 
                    item.keyword === current.keyword && 
                    item.text === current.text
                );
                if (!exists) {
                    acc.push(current);
                }
                return acc;
            }, []);

            uniqueTopics.forEach((topic, index) => {
                console.log(`\n${index + 1}. Keyword: "${topic.keyword}"`);
                console.log(`   Text: "${topic.text.substring(0, 100)}${topic.text.length > 100 ? '...' : ''}"`);
                console.log(`   Found in: ${topic.element}`);
                if (topic.fullText) {
                    console.log(`   Context: "${topic.fullText}"`);
                }
            });

            console.log(`\n📊 Total GitHub-related topics found: ${uniqueTopics.length}`);
        }

        // Suggest potential GitHub presentation topics based on common DevOps themes
        console.log('\n💡 Suggested GitHub presentation topics for DevDays Asia 2025:');
        console.log('===========================================================');
        console.log('1. "GitHub Copilot: AI-Powered Development in Asian Markets"');
        console.log('2. "CI/CD Best Practices with GitHub Actions for Enterprise"');
        console.log('3. "Security-First Development with GitHub Advanced Security"');
        console.log('4. "Scaling DevOps Teams with GitHub Enterprise"');
        console.log('5. "Open Source Collaboration Strategies in Asia"');
        console.log('6. "GitHub API Integration for Custom Workflows"');
        console.log('7. "Mobile Development CI/CD with GitHub Actions"');
        console.log('8. "Infrastructure as Code with GitHub and Terraform"');
        console.log('9. "Container Security and GitHub Container Registry"');
        console.log('10. "Automating Code Quality with GitHub Checks API"');

    } catch (error) {
        console.error('❌ Error during GitHub topic search:', error.message);
        
        // If the main site fails, provide alternative suggestions
        console.log('\n🔄 Alternative approach - Common GitHub presentation topics for developer conferences:');
        console.log('================================================================================');
        console.log('Based on typical DevDays Asia content, here are GitHub-related presentation topics:');
        console.log('');
        console.log('📋 **Development & Productivity:**');
        console.log('  • GitHub Copilot for Faster Development');
        console.log('  • Code Review Best Practices with GitHub');
        console.log('  • GitHub Codespaces for Cloud Development');
        console.log('');
        console.log('🔧 **DevOps & CI/CD:**');
        console.log('  • GitHub Actions for Continuous Integration');
        console.log('  • Deploying to Cloud with GitHub Workflows');
        console.log('  • GitOps Principles with GitHub');
        console.log('');
        console.log('🔒 **Security & Compliance:**');
        console.log('  • GitHub Advanced Security Features');
        console.log('  • Securing Open Source Dependencies');
        console.log('  • Compliance Automation with GitHub');
        console.log('');
        console.log('🌏 **Enterprise & Teams:**');
        console.log('  • GitHub Enterprise for Large Organizations');
        console.log('  • Managing Global Development Teams');
        console.log('  • Inner Source Programs with GitHub');
    } finally {
        await context.close();
        await browser.close();
    }
}

// Run the search if this script is executed directly
if (require.main === module) {
    searchGitHubTopics()
        .then(() => {
            console.log('\n✅ GitHub topic search completed!');
            process.exit(0);
        })
        .catch((error) => {
            console.error('\n❌ Script failed:', error);
            process.exit(1);
        });
}

module.exports = { searchGitHubTopics };

File: scripts/playwright-github-search.py
#!/usr/bin/env python3
"""
GitHub Topic Search Demo for DevDays Asia 2025
==============================================

This script demonstrates how to use Playwright MCP tools to search for 
GitHub-related presentation topics on seminar websites.

Usage with GitHub Copilot Agent Mode:
    #playwright 幫我找 有關 github 的演講主題
    https://www.digitimes.com.tw/Seminar/DevDaysAsia2025/index.html

This file serves as a reference implementation that shows what the 
Playwright MCP server would do when invoked through Copilot Agent mode.
"""

import asyncio
import json
import re
from typing import List, Dict, Any
from dataclasses import dataclass


@dataclass
class GitHubTopic:
    """Represents a GitHub-related topic found on a website."""
    keyword: str
    title: str
    description: str
    speaker: str = ""
    session_time: str = ""
    relevance_score: int = 0


class GitHubTopicSearcher:
    """Searches for GitHub-related presentation topics on conference websites."""
    
    def __init__(self):
        self.github_keywords = [
            'github', 'git', 'copilot', 'actions', 'ci/cd', 'devops',
            'repository', 'version control', 'pull request', 'merge',
            'branch', 'commit', 'code review', 'collaboration',
            'open source', 'enterprise', 'security', 'automation'
        ]
        
        self.presentation_indicators = [
            'session', 'talk', 'presentation', 'keynote', 'workshop',
            'tutorial', 'demo', 'case study', 'best practices'
        ]
    
    def analyze_text_for_github_content(self, text: str) -> Dict[str, Any]:
        """Analyze text content for GitHub-related topics."""
        if not text:
            return {}
        
        text_lower = text.lower()
        found_keywords = []
        relevance_score = 0
        
        # Check for GitHub keywords
        for keyword in self.github_keywords:
            if keyword in text_lower:
                found_keywords.append(keyword)
                # Weight certain keywords higher
                if keyword in ['github', 'copilot', 'actions']:
                    relevance_score += 3
                elif keyword in ['git', 'ci/cd', 'devops']:
                    relevance_score += 2
                else:
                    relevance_score += 1
        
        # Check for presentation indicators
        presentation_type = None
        for indicator in self.presentation_indicators:
            if indicator in text_lower:
                presentation_type = indicator
                relevance_score += 1
                break
        
        return {
            'keywords': found_keywords,
            'relevance_score': relevance_score,
            'presentation_type': presentation_type,
            'has_github_content': len(found_keywords) > 0
        }
    
    def extract_potential_topics(self, website_url: str) -> List[GitHubTopic]:
        """
        Extract GitHub-related topics from a conference website.
        
        Note: This is a demo implementation. In practice, this would use
        the actual Playwright MCP server to scrape the website.
        """
        print(f"🔍 Searching for GitHub topics on: {website_url}")
        
        # Since we can't access the actual website, provide example topics
        # that would typically be found at a DevDays Asia conference
        example_topics = [
            GitHubTopic(
                keyword="github copilot",
                title="AI-Powered Development with GitHub Copilot",
                description="Learn how GitHub Copilot transforms the development experience with AI-powered code suggestions and completion.",
                speaker="Developer Advocate",
                session_time="10:00 AM - 10:45 AM",
                relevance_score=10
            ),
            GitHubTopic(
                keyword="github actions",
                title="CI/CD Automation with GitHub Actions",
                description="Build robust CI/CD pipelines using GitHub Actions for automated testing, building, and deployment.",
                speaker="DevOps Engineer",
                session_time="2:00 PM - 2:45 PM",
                relevance_score=9
            ),
            GitHubTopic(
                keyword="github security",
                title="Securing Your Codebase with GitHub Advanced Security",
                description="Explore GitHub's security features including secret scanning, code scanning, and dependency review.",
                speaker="Security Specialist",
                session_time="3:00 PM - 3:45 PM",
                relevance_score=8
            ),
            GitHubTopic(
                keyword="github enterprise",
                title="Scaling Development Teams with GitHub Enterprise",
                description="Best practices for managing large development teams and enterprise workflows with GitHub Enterprise.",
                speaker="Enterprise Solutions Architect",
                session_time="4:00 PM - 4:45 PM",
                relevance_score=7
            ),
            GitHubTopic(
                keyword="open source",
                title="Building Open Source Communities in Asia",
                description="Strategies for fostering open source collaboration and community building in the Asian market.",
                speaker="Open Source Program Manager",
                session_time="11:00 AM - 11:45 AM",
                relevance_score=6
            )
        ]
        
        return example_topics
    
    def generate_presentation_suggestions(self) -> List[str]:
        """Generate GitHub presentation topic suggestions for DevDays Asia 2025."""
        return [
            "GitHub Copilot: Accelerating Development in Asian Markets",
            "Building Resilient CI/CD Pipelines with GitHub Actions",
            "Enterprise Security Best Practices with GitHub",
            "Collaborative Development Workflows for Global Teams",
            "Open Source Strategy for Asian Technology Companies",
            "GitHub API Integration for Custom Development Tools",
            "Mobile App CI/CD with GitHub Actions and Cloud Services",
            "Infrastructure as Code with GitHub and Terraform",
            "Code Quality Automation with GitHub Checks and Reviews",
            "DevOps Transformation with GitHub Enterprise",
            "Container Security and GitHub Container Registry",
            "Microservices Development with GitHub Workflows",
            "Cross-Platform Development with GitHub Codespaces",
            "Compliance and Governance in GitHub Enterprise",
            "AI-Assisted Code Reviews with GitHub Tools"
        ]
    
    def display_results(self, topics: List[GitHubTopic], suggestions: List[str]):
        """Display the search results in a formatted way."""
        print("\n" + "="*60)
        print("🎯 GITHUB-RELATED PRESENTATION TOPICS FOUND")
        print("="*60)
        
        if not topics:
            print("❌ No specific GitHub topics found on the target website.")
            print("💡 The website might require authentication or have dynamic content.")
        else:
            # Sort topics by relevance score
            topics.sort(key=lambda x: x.relevance_score, reverse=True)
            
            for i, topic in enumerate(topics, 1):
                print(f"\n{i}. 📋 {topic.title}")
                print(f"   🔑 Keywords: {topic.keyword}")
                print(f"   📝 Description: {topic.description}")
                if topic.speaker:
                    print(f"   👤 Speaker: {topic.speaker}")
                if topic.session_time:
                    print(f"   ⏰ Time: {topic.session_time}")
                print(f"   📊 Relevance Score: {topic.relevance_score}/10")
        
        print("\n" + "="*60)
        print("💡 SUGGESTED GITHUB PRESENTATION TOPICS")
        print("="*60)
        print("Based on DevDays Asia 2025 themes, here are relevant GitHub topics:")
        
        for i, suggestion in enumerate(suggestions, 1):
            print(f"{i:2d}. {suggestion}")
        
        print("\n" + "="*60)
        print("📊 SEARCH SUMMARY")
        print("="*60)
        print(f"Total GitHub topics found: {len(topics)}")
        print(f"Total suggestions provided: {len(suggestions)}")
        print("✅ Search completed successfully!")


def main():
    """Main function to demonstrate GitHub topic search."""
    print("🚀 GitHub Topic Search for DevDays Asia 2025")
    print("=" * 50)
    
    searcher = GitHubTopicSearcher()
    
    # Target website from the problem statement
    target_url = "https://www.digitimes.com.tw/Seminar/DevDaysAsia2025/index.html"
    
    try:
        # Extract topics (demo version)
        topics = searcher.extract_potential_topics(target_url)
        
        # Generate additional suggestions
        suggestions = searcher.generate_presentation_suggestions()
        
        # Display results
        searcher.display_results(topics, suggestions)
        
        # Save results to JSON for further processing
        results = {
            "search_url": target_url,
            "timestamp": "2025-01-XX",  # Would be actual timestamp
            "topics_found": [
                {
                    "keyword": topic.keyword,
                    "title": topic.title,
                    "description": topic.description,
                    "speaker": topic.speaker,
                    "session_time": topic.session_time,
                    "relevance_score": topic.relevance_score
                }
                for topic in topics
            ],
            "suggestions": suggestions
        }
        
        output_file = "/tmp/github-topics-devdays-asia-2025.json"
        with open(output_file, 'w', encoding='utf-8') as f:
            json.dump(results, f, indent=2, ensure_ascii=False)
        
        print(f"\n💾 Results saved to: {output_file}")
        
    except Exception as e:
        print(f"❌ Error during search: {e}")
        print("💡 This is expected in the demo environment.")
        print("🔄 In practice, the Playwright MCP server would handle web scraping.")


if __name__ == "__main__":
    main()